### PR TITLE
tth 743 updates

### PR DESCRIPTION
**Pull request by jpata**
This is the vhbb side of the sync updates from tth, up for discussion.

- GenHFHadronMatcher to label heavy flavour (eg. tt+b vs tt+B) in preprocessor
- GenJets re-clustered due to GenHFHadronMatcher/miniAOD mismatch, saved with hadron matching info
- some jet, lepton variables for sync
- JEC and uncertainty evaluated, attached to jet
- event b-tag uncertainty evaluated using CSV weights

I've tested it under 722 and ran crab jobs of the ttH reference analysis exercise.

NOTE: currently the GenHFHadronMatcher recipe relies on getting a bugfix manually from the twiki for the plugin, I don't know of a good way to put it into this repo while the matcher guys succeed in fixing it in the release. The fix is the following.
~~~
diff --git a/PhysicsTools/JetMCAlgos/plugins/GenHFHadronMatcher.cc b/PhysicsTools/JetMCAlgos/plugins/GenHFHadronMatcher.cc
index 9c6ec2b..578fa93 100644
--- a/PhysicsTools/JetMCAlgos/plugins/GenHFHadronMatcher.cc
+++ b/PhysicsTools/JetMCAlgos/plugins/GenHFHadronMatcher.cc
@@ -384,8 +384,8 @@ std::vector<int> GenHFHadronMatcher::findHadronJets ( const reco::GenParticleCol
         std::vector <int> LastQuarkMotherId;

         int hadFlav = hadMothers.at(hadIdx).pdgId() <0?-1:1; // Charge of the hadron (-1,1)
-        if ( abs ( hadMothers.at(hadIdx).pdgId() ) /1000 < 1 ) {
-            hadFlav*=-1;    // Inverting flavour of hadron if it is a meson
+        if ( abs ( hadMothers.at(hadIdx).pdgId() ) /1000 != 5 && abs ( hadMothers.at(hadIdx).pdgId() ) /100%10 == 5 ) {
+            hadFlav*=-1;    // Inverting flavour of B hadron if it is a meson
         }

         // Searching only first quark in the chain with the same flavour as hadron
@@ -738,8 +738,8 @@ int GenHFHadronMatcher::findInMothers ( int idx, std::vector<int> &mothChains, s
     int foundStopId = -1;
     int pdg_1 = hadMothers.at ( idx ).pdgId();
     int partCharge = ( hadMothers.at ( idx ).pdgId() >0 ) ?1:-1;
-// Inverting charge if mother is a b(c) meson
-    if ( abs ( hadMothers.at ( idx ).pdgId() ) /1000 < 1 && ( abs ( hadMothers.at ( idx ).pdgId() ) /100%10 == 4 || abs ( hadMothers.at ( idx ).pdgId() ) /100%10 == 5 ) ) {
+    // Inverting charge if mother is a b meson
+    if ( abs ( hadMothers.at ( idx ).pdgId() ) /1000 != 5 && abs ( hadMothers.at ( idx ).pdgId() ) /100%10 == 5 ) {
         partCharge*=-1;
     }
~~~

@bianchini @gkasieczka @arizzi

### CODE REVIEW

File: VHbbAnalysis/Heppy/test/combined_cmssw.py [hunk 1]
```diff
@@ -144,6 +144,62 @@
     trimPtFracMin = cms.double(0.06),
     useExplicitGhosts = cms.bool(True))
 
+###
+### GenHFHadronMatcher
+###
+process.load("SimGeneral.HepPDTESSource.pythiapdt_cfi")
```
Comment (arizzi) [inline VHbbAnalysis/Heppy/test/combined_cmssw.py]: how would this work on realdata?

Comment (jpata) [inline VHbbAnalysis/Heppy/test/combined_cmssw.py]: right, there should be an isMC around it and also the genjet clustering.

**Comment by arizzi** [comment]
concerning the fix above you can just
git cms-addpkg PhysicsTools/JetMCAlgos
# apply the fix
git add ...theChangedFiles..
git commit
git push

what's the problem with this?

**Comment by jpata** [comment]
then the vhbb PR will contain changes to PhysicsTools/JetMCAlgos - I thought it is bad if we modify stuff which is not ours.

**Comment by arizzi** [comment]
it is, but is still better than having to cut and paste the diffs... then when the fix is coming we will merge it

**Comment by jpata** [comment]
ok. shall I add this to PR #67 or #68 (heppy side against cbernet)

**Comment by arizzi** [comment]
just put the fix in the Vhbb side (or the lack of the fix would make #68
unusable in standard heppy?)

**Comment by jpata** [comment]
right, it belongs in #67, nothing about genHF in #68. without the fix #67 is usable but the hadron matcher produces subtly wrong results.

**Comment by arizzi** [comment]
can I merge #67 and #68 as they are or you want to add more stuff here? I'd say to make a new PR for other features and I merge these... @jpata let me know

**Comment by jpata** [comment]
#67 please wait one more hour, I'm just finishing the per-jet weights. #68 I don't change any more.

**Comment by arizzi** [comment]
ok, for 68 if colin merges your new PR I'd rather merge from that side

**Comment by jpata** [comment]
OK fine to merge #67, thanks.

**Comment by arizzi** [comment]
new branch is vhbbHeppy745 see #75

**Comment by jpata** [comment]
@arizzi, I'm very sorry, I thought what I did was clear from the code and my original commit msg. I should have been more clear. I'll be back online in a few days and I can take care of this as you see fit.
> GenJets re-clustered due to GenHFHadronMatcher/miniAOD mismatch, saved with hadron matching info

**Comment by arizzi** [comment]
do you have a new PR? how do we want to handle this?

**Comment by jpata** [comment]
@arizzi just so that I'm on the same page: this was merged for V12 in vhbb, but it is preferred to have an additional GenJet collection over replacing. (this may no longer be necessary, as this was some "bug" in the hadron matcher module)
I can clean up for the latest HEAD and do necessary changes as you see fit.

**Comment by arizzi** [comment]
those are the open points in my view:
- we should not remake the gen jets as they should be the same but I understand that for the hadron matching this could be actually needed? can you confirm?
- in output we should have a single gen jet having all needed features, namely:
   * the standard flavour coming from PAT
   * the hadron matched flavour
   * the genNu correction

if hadron matcher needs to remake its own jets, then we have to back match those to the original collection and just set a had-flavour attribute

**Pull request closed without merging**